### DISCUSSION TITLE
Ctrl-C in Dagger CLI doesn't propagate to service

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12371] [Help]
## Problem

When running a service from the Dagger CLI, for example `dagger call .... up`, interrupting the call with Ctrl-C does not propagate the SIGINT signal to the service process itself. As a result, the program doesn't get a chance to clean up.

This may also apply to `WithExec`, I haven't checked.

**Comment by jedevc** [comment]
Hm, this is tricky. cc @vito :tada:

I'm not *necessarily* convinced that we should propagate a SIGINT through to services being run - the processes are being run remotely, etc. However, I do agree that running services should get a graceful shutdown if possible using SIGTERM.

This is what I think makes most sense (this bit at least seems easy):
- We need the dagger cli to intercept SIGINT/SIGTERM, and use that to try to shut down services that are running from `dagger call ... up` (we can use #6354 as a basis)
	- Specifically, the `dagger session` *shouldn't* get these auto-magically (which is the case today through process groups).
- If we send multiple SIGINT/SIGTERM, then we don't graceful stop, and kill the service.

A couple open questions:
- Is this just for zenith modules? What about running the SDK directly? *Should* that immediately take down the session (because of process groups), or should the SDK be responsible for clearing up (and are users responsible for writing that code)?
- Does this also apply for services started automatically with `WithServiceBinding`? We've been assuming that if you care about shutdown, you need to manually call `Start`/`Stop`, but that works less well if you're returning services from a module function (you can't call `Stop` after returning)?

> This may also apply to WithExec, I haven't checked.

Hm, I'm not convinced this should matter - the process stops, but obviously this is never cached, etc, so the result is just discarded (and we re-run it on next time).

**Comment by jedevc** [comment]
Linking this to https://github.com/dagger/dagger/issues/6493, since there's continued discussion of service stop semantics there as well.

**Comment by levlaz** [comment]
This bit me a bit too, if I have a CLI that is struggling to connect to an engine there is no way for me to kill it, I tried Ctrl+c Ctrl+z and even whipped out `Enter ~ .` (which seemed to help back in they day where I got stuck in serial consoles) and nothing works.

**Comment by levlaz** [comment]
This has been fixed in 0.11! 😎

**Comment by helderco** [comment]
> This has been fixed in 0.11! 😎

Closing then! Re-open if the issue persists.

**Comment by MikaelElkiaer** [comment]
@shykes @helderco I am possibly experiencing this issue for `WithExec`, using `v0.11.4`.
It might also be that I am misunderstanding something.
I made a small example here: https://github.com/MikaelElkiaer/daggerverse/blob/main/testing.go

I would like to have the interrupt propagated to the underlying process of the `WithExec`.
Concretely, in another [function](https://github.com/MikaelElkiaer/daggerverse/blob/8c84fb26a6408e7ed2369e9a79145b6db0ece1b4/helm.go#L150), I am trying to run `helm install --atomic <...>` and rely on the cleanup which follows a timeout or interrupt.

Edit: Update link, ping people

**Comment by MikaelElkiaer** [comment]
@shykes @helderco This is still an issue with `v0.12.7`.
